### PR TITLE
dkms: update to 3.0.6 and update other related packages

### PR DESCRIPTION
**Pull request by subnut**
#### Testing the changes
- I tested the changes in this PR: **YES**

#### Local build testing
- I built this PR locally for my native architecture, x86_64-glibc

### CODE REVIEW

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [inline srcpkgs/xbps-triggers/files/dkms:42]: For a demo of this code
- Ensure you are on linux `5.15` or greater
- Install `ddcci-dkms-0.4.1` (from any commit before #37127 is merged)
  (During installation, you should see this message)
  ```
  ddcci-dkms-0.4.1_1: configuring ...
  DKMS module 'ddcci-0.4.1' already added, skipping.
  Building DKMS module 'ddcci-0.4.1' for kernel-5.xx.yy_z... FAILED!
  DKMS module 'ddcci-0.4.1' failed to build, please check /var/lib/dkms
  for errors in the log file.
  ```
- Run `xbps-remove ddcci-dkms`
  (You should see this line in the output)
  ```
  Removing DKMS module 'ddcci-0.4.1' for kernel-5.xx.yy_z... FAILED!
  ```

File: srcpkgs/broadcom-wl-dkms/template
Comment (ahesford) [inline srcpkgs/broadcom-wl-dkms/template:17]: This is unnecessary. Almost no users will require that this driver be present in the initramfs, and those that do probably have custom setups to deal with the complexities of bringing up a wireless interface in an early-boot environment anyway. Better to let those users deal with manually updating their images than force every user of this driver to regenerate *all* initramfs images with an update.

File: srcpkgs/dkms/template
Comment (ahesford) [inline srcpkgs/dkms/template:27]: Just keep the version we have in `$FILESDIR`. Our script is a bit more verbose (listing the kernel package name) and explicitly calling `/usr/bin/dkms` is better than calling `dkms` when the script verifies that the former exists but the latter is subject to `$PATH` shenanigans.

This will also make the script less likely to subtly break in a future update if there are unnoticed changes.

File: srcpkgs/xbps-triggers/files/dkms
Comment (ahesford) [inline srcpkgs/xbps-triggers/files/dkms:41]: What problem does this fix?

`rm -rf "/var/lib/dkms/$VAR"` is pretty dangerous when the vars haven't been carefully sanitized. If there's ever a format change in the DKMS output, this could break horribly.

**Review by ahesford — CHANGES_REQUESTED** [review]

File: srcpkgs/broadcom-wl-dkms/template
Comment (subnut) [reply srcpkgs/broadcom-wl-dkms/template:17]: Removed.

File: srcpkgs/dkms/template
Comment (subnut) [reply srcpkgs/dkms/template:27]: Done. I've also made our script a bit more efficient, and also POSIX-compatible.

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms:41]: > What problem does this fix?

See this - https://github.com/void-linux/void-packages/pull/37139#discussion_r873132092

> `rm -rf "/var/lib/dkms/$VAR"` is pretty dangerous

I agree.  

Also, after looking into what causes it, this looks like a `dkms` problem. So, I should probably leave it to them to fix this. Not our job.

_Removed._

Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms]: #### UPDATE:
Nope. This was an bug in the dkms trigger. But this bug existed since a long time ago. Not introduced by this PR.

I've fixed it in the latest commit, and also explained the bug in the commit message.

File: srcpkgs/xbps-triggers/files/dkms
Comment (ahesford) [inline srcpkgs/xbps-triggers/files/dkms:34]: This breaks version parsing for old versions of DKMS, which we can't do in case somebody does a partial system update. Suggest instead:
```sh
$DKMS status -m "$1" | while IFS="/, " read _modname _modver _line; do
	_modver="${_modver%:}"
	shift 2
```
which *should* reliably separate `$_modname` and `$_modver` in both old and new DKMS, leaving the rest of the line intact. Note that we still strip the possible `:` suffix in the version here.

File: srcpkgs/xbps-triggers/files/dkms
Comment (ahesford) [inline srcpkgs/xbps-triggers/files/dkms:47]:
```suggestion
			if [ "${_line}" = added ]; then
				# The module wasn't built successfully for any kernel version
				echo -n "Cleaning up DKMS module '${_modname}-${_modver}'... "
				$DKMS remove -m "${_modname}" -v "${_modver}" >/dev/null 2>&1
			else
				echo -n "Removing DKMS module '${_modname}-${_modver}' for kernel-${_kver}... "
				$DKMS remove -m "${_modname}" -v "${_modver}" -k "${_kver}" >/dev/null 2>&1
			fi
```

**Review by ahesford — CHANGES_REQUESTED** [review]

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms:47]: Okay.

But what if `$1` and `$_modname` are different. Should we check for that? How should we handle that?

Comment (ahesford) [reply srcpkgs/xbps-triggers/files/dkms:47]: The trigger already fails to verify that the module name reported by DKMS, so let's not worry about it. If you really think this could be an issue, verify that `"$1" = "${_modname}"` or skip the iteration.

My other suggestion had two bugs: 1. the shift shouldn't be applied because it is already done outside the loop (this is also a bug in the existing `shift 2; continue` line); 2. I'm not setting the kernel version properly. I think the proper solution is something more like

```sh
remove_modules() {
	local _modname _modver _kver _line _arch _status

	# Remove the specified modules from all kernels.
	set -- ${dkms_modules}
	while [ $# -gt 0 ]; do
		$DKMS status -m "$1" | while IFS="/, " read _modname _modver _line; do
			[ "$1" = "${_modname}" ] || continue

			_modver="${_modver%:}"

			if [ "${_line}" = added ]; then
				# The module wasn't built successfully for any kernel version
				echo -n "Cleaning up DKMS module '${_modname}-${_modver}'... "
				$DKMS remove -m "${_modname}" -v "${_modver}" >/dev/null 2>&1
				continue
			fi

			read _kver _arch _status <<-EOF
			${_line}
			EOF

			case "${_status}" in
				added|built|installed) ;;
				*) break ;;
			esac

			echo -n "Removing DKMS module '${_modname}-${_modver}' for kernel-${_kver}... "
			$DKMS remove -m "${_modname}" -v "${_modver}" -k "${_kver}" >/dev/null 2>&1
                done
                shift 2
        done
}


```

File: srcpkgs/xbps-triggers/files/dkms
Comment (ahesford) [inline srcpkgs/xbps-triggers/files/dkms:29]: `IFS=':/, '` makes the cut a bit more aggressive than my earlier proposal, but this is probably fine. The two-stage cut in my proposal restricted the more generic `IFS='/, '` cut to only two fields (module name and version), then applied the more specific `IFS=', '` cut to only the remainder. Is there any legitimate concern that any of the fields might have characters ':' or '/' that would foil this split?

File: srcpkgs/xbps-triggers/files/dkms
Comment (ahesford) [inline srcpkgs/xbps-triggers/files/dkms:36]: These two clauses should be flipped because `"$kver" = added` is more specific. Also, it might be worthwhile to verify that `$arch` and `$status` are empty, or the response isn't in the expected format.

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms]: Good point. I'd never thought if the `:` and `,` characters might appear in a module's name.

None of the fields can have the `/` character, as that would break dkms itself. I've verified that.

I'm more concerned about the `:` and `,` characters. They have the possibility to break this script.

I've pushed a fix in the latest commit.

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms]: done.

**Comment by subnut** [comment]
Rebased on `master`.

**Comment by subnut** [comment]
@ahesford Have you reviewed the updated logic? (I've now used `tr` to make the new output similar to the old one)

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms]: **UPDATE:**
No, there is no concern that `,` `:` `/` characters might appear in any of the fields.  
`dkms` itself assumes so.

https://github.com/dell/dkms/blob/ea97ad27efb4d227242e322f9d30c7280a6eba47/kernel_prerm.d_dkms#L7-L11

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms]: Your above example doesn't print `done.` or `FAILED!` after the `... ` to indicated whether the action succeeded or failed.

**Comment by subnut** [comment]
Sorry for the delay, I am busy.

---

@ahesford It makes the logic simpler IMO.  

I've updated the logic again, and this time I've only utilized the expected format instead of using workarounds like `$kver = added`. Hope this alleviates your concerns.

If you have any better suggestions, do share. I'll implement it ASAP.

**Comment by subnut** [comment]
Reference for `IFS` value -  
https://github.com/dell/dkms/blob/ea97ad27efb4d227242e322f9d30c7280a6eba47/kernel_prerm.d_dkms#L7-L11

File: srcpkgs/zfs/patches/0001-only-build-the-module-in-dkms.conf.patch
Comment (subnut) [reply srcpkgs/zfs/patches/0001-only-build-the-module-in-dkms.conf.patch:35]: No, it is not a hard error yet.

Comment (subnut) [reply srcpkgs/zfs/patches/0001-only-build-the-module-in-dkms.conf.patch]: Removed.

File: srcpkgs/xbps-triggers/files/dkms
Comment (subnut) [reply srcpkgs/xbps-triggers/files/dkms]: Done.

**Review by ahesford — COMMENTED** [review]
This seems acceptable, but we need to test thoroughly. Accidentally breaking DKMS with trigger and hook changes would be bad news.

**Comment by subnut** [comment]
Updated to latest release and rebased on `master`.

File: srcpkgs/dkms/template
Comment (ahesford) [inline srcpkgs/dkms/template:6]: Please add a dependency on `xbps-triggers>=0.123_1` to force updates to pull the new xbps-triggers, otherwise the trigger will be broken should people attempt a partial update. After that, this is ready for merge.

**Review by ahesford — CHANGES_REQUESTED** [review]

File: srcpkgs/dkms/files/kernel.d/dkms.prerm
Comment (ahesford) [inline srcpkgs/dkms/files/kernel.d/dkms.prerm:8]: `IFS` needs to accommodate whitespace, too. See comment in the `xbps-trigger` script.

File: srcpkgs/xbps-triggers/files/dkms
Comment (ahesford) [inline srcpkgs/xbps-triggers/files/dkms:28]: On further testing, it turns out `IFS=,:/` is not sufficient because it doesn't recognize whitespace as a separator. As a consequence, `$kver`, `$arch` and `$status` all contain leading spaces. That causes the test on line 35 to fail and modules are never removed when the package is removed.

Try, for example, with this version of `xbps-triggers` and `dkms` installed along with one or more kernels:
```sh
dkms status
xbps-install nvidia-dkms
dkms status
xbps-remove nvidia-dkms
dkms status
```

The output of `xbps-remove` will show `Removing nvidia-dkms-515.57_1... ` and regenerate your initramfs images, but it will not print any messages about removing the module for specific kernels. The final `dkms status` will report an error and provide an incomplete listing of installed modules because it chokes on `nvidia` after the package files are removed.

**Review by ahesford — CHANGES_REQUESTED** [review]
Welp, this isn't ready to go after all. The trigger and kernel hook need some fixes.

File: srcpkgs/dkms/template
Comment (subnut) [reply srcpkgs/dkms/template]: Done.

**Comment by subnut** [comment]
Sorry for the `IFS` blunder :sweat_smile:  
Life's been very hectic recently.

**Comment by subnut** [comment]
After searching in leahneukirchen's https://inbox.vuxu.org/voidlinux-github, I've found how the `IFS` value got wrong.

- [This][1] is the last patch where the `IFS` is set correctly
- [This][2] is the first patch without `IFS`
- [This][3] is the last patch without `IFS`
- [This][4] is the first patch where `IFS` is set to the wrong value

[1]: https://inbox.vuxu.org/voidlinux-github/20220518173718.-EX16uUeKbHWxd3cdsNdaUz-WLTr1cou3FmmR8hS-eM@z/
[2]: https://inbox.vuxu.org/voidlinux-github/20220521040840.xP7XR634VvrByVkSU5hcjoXoMhv9peu5qnXsI39waMw@z/
[3]: https://inbox.vuxu.org/voidlinux-github/20220524021940.wCfmV3TNx79wHGiLEv7ytRRl6o9hRGLyapoKhObjP6U@z/
[4]: https://inbox.vuxu.org/voidlinux-github/20220622070401.3ZCiPp-cMiimkXqvRRh-CUtc4YtBmSy7Q04JXldVDwg@z/

Just for historical purposes.


<br>
<br>


_Now_ I understand why some projects prefer Gerrit. :laughing:

**Comment by dataCobra** [comment]
Hey @subnut,

is there something I can help you with?

**Pull request closed without merging**
